data_steward/cdr_cleaner/cleaning_rules/sandbox_and_remove_pids.py

```diff
@@ -0,0 +1,109 @@
+# Project imports
+from constants.cdr_cleaner import clean_cdr as clean_consts
+import constants.cdr_cleaner.clean_cdr as cdr_consts
+from notebooks import bq
+from constants import bq_utils as bq_consts
+from sandbox import get_sandbox_dataset_id
+
+# Query to create tables in sandbox with rows that will be removed per cleaning rule
+SANDBOX_QUERY = """
+CREATE OR REPLACE TABLE `{project}.{sandbox_dataset}.{intermediary_table}` AS (
+SELECT *
+FROM `{project}.{dataset}.{table}`
+WHERE person_id IN({pids}))
+"""
+
+# Query to truncate existing tables to remove PIDs based on cleaning rule criteria
+CLEAN_QUERY = """
+SELECT *
+FROM `{project}.{dataset}.{table}`
+WHERE person_id NOT IN({pids})
+"""
+
+# Query to list all tables within a dataset that contains person_id in the schema
+PERSON_TABLE_QUERY = """
+SELECT table_name
+FROM `{project}.{dataset}.INFORMATION_SCHEMA.COLUMNS`
+WHERE COLUMN_NAME = 'person_id'
+"""
+
+
+def get_tables_with_person_id(project_id, dataset_id):
+    """
+    Get list of tables that have a person_id column, excluding mapping tables
+    """
+    person_table_query = PERSON_TABLE_QUERY.format(project=project_id,
+                                                   dataset=dataset_id)
+    person_table_list = list(
+        bq.query(person_table_query).table_name.get_values())
+
+    # exclude mapping tables from list, to be removed after all cleaning rules
+    for item in person_table_list:
+        if item.startswith('_mapping'):
+            person_table_list.remove(item)
+
+    return person_table_list
+
+
+def get_sandbox_queries(project_id, dataset_id, pids_query, ticket_number):
+    """
+    Returns a list of queries of all tables to be added to the datasets sandbox. These tables include all rows from all
+    effected tables that include PIDs that will be removed by a specific cleaning rule.
+
+    :param project_id: bq project_id
+    :param dataset_id: bq dataset_id
+    :param pids_query: query from cleaning rule that lists all PIDs that need to be sandboxed and removed
+    :param ticket_number: ticket number from jira that will be appended to the end of the sandbox table names
+    :return: list of CREATE OR REPLACE queries to create tables in sandbox
+    """
+    person_tables_list = get_tables_with_person_id(project_id, dataset_id)
+    # take pids_query and generate list of pids
+    pid_list = bq.query(
```

Comment:
It might be a good idea to move `pid_list = bq.query(` out of this function for better separation of function units and easier testing. The purpose of this function is to sandbox all records associated with a list of `person_ids`, but in the current implementation, it does two things (run `pids_query` to dynamically retrieve `person_ids` and sandbox records), we might want to further separate these two basic function units.   

It would be easier for writing tests, for instance, in the integration test for `get_sandbox_queries`, we don't want to bother with the correctness of `pids_query`, we just want to pass in a list of `person_ids`. Other functions should be responsible for running `pids_query`. I would suggest moving `pid_list = bq.query(` to your own cleaning rule. 

Another reason that this might limit us is that we assume`pids_query` is always rendered the same way using `project_id` and `dataset_id` in `pids_query.format(project=project_id,dataset=dataset_id)`, but it may not be the case (e.g. in the cleaning rule I am working on), we don't want to limit ourselves to the way the `pids_query` is rendered and and this function probably should not how the `pids_query` is rendered or constructed.

Reply:
Sounds good, I am going to remove the functionality the `pids_query` and replace the parameter as a list of PIDS that we pass from each separate cleaning rule.